### PR TITLE
AP-167

### PR DESCRIPTION
**Pull request by ptravers**
This PR adds a method to construct a S3File for a directory upload

### CODE REVIEW

File: src/bf/model/upload.rs [hunk 1]
```diff
@@ -239,10 +241,28 @@ impl S3File {
...
+                        None => None
+                    }
+                })
+                .map(|path| path.to_owned());
+
         // And the resulting metadata so we can pull the file size:
         let metadata = fs::metadata(file_path)?;
 
-        Ok((file_name, metadata))
+        println!("{:?}", destination_path);
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: stray println

File: src/bf/model/upload.rs [hunk 2]
```diff
@@ -239,10 +241,28 @@ impl S3File {
 
         let file_name = file_name?;
 
+        let canonical_dir_path = directory_path.canonicalize()?;
+
+        let destination_path: Option<String> = 
+            file_path
+                .strip_prefix(canonical_dir_path)
+                .ok()
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: I think this should always succeed, and if it doesn't then there is a larger bug in here. So I think it makes sense to bubble that error up `file_path.strip_prefix(canonical_dir_path)?;`

File: src/bf/model/upload.rs [hunk 3]
```diff
@@ -239,10 +241,28 @@ impl S3File {
...
+                .and_then(|path| path.parent())
+                .and_then(|path| {
+                    match path.to_str() {
+                        Some("") => None,
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: what does a file_path = `None` value imply? Shouldn't every file have a path?

File: src/bf/model/upload.rs [hunk 4]
```diff
@@ -298,6 +348,7 @@ impl S3File {
                 total_chunks: (self.size as f64 / c as f64).floor() as usize + 1,
             }),
             multipart_upload_id: self.multipart_upload_id,
+            file_path: None,
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: self.file_path

File: src/bf/model/upload.rs [hunk 5]
```diff
@@ -312,6 +363,7 @@ impl S3File {
             size: self.size,
             chunked_upload: self.chunked_upload,
             multipart_upload_id,
+            file_path: None,
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: same here, we should be able to pass this value through

File: src/bf/model/upload.rs [hunk 6]
```diff
@@ -599,4 +651,40 @@ mod tests {
...
+    #[test]
+    pub fn during_directory_upload_root_upload_directory_path_finding_works() {
+        let path = concat!(env!("CARGO_MANIFEST_DIR"), "/test/data/").to_owned();
+        let file = concat!(env!("CARGO_MANIFEST_DIR"), "/test/data/small/example.csv").to_owned();
+
+        let result = S3File::retaining_file_path(file, path, None);
+
+        match result {
+            Err(err) => panic!("failed to get directory {:?}", err),
+            Ok(_) => {
```
Comment (mattusifer) [inline src/bf/model/upload.rs]: this could be written `Ok(s3_file)` and then you won't have to unwrap below.

File: src/bf/model/upload.rs (hunk 1 above)
Comment (ptravers) [reply src/bf/model/upload.rs]: yeah already fixed that guy

**Comment by ptravers** [comment]
@mattusifer yeah so the None represents root in upload/preview. We can always send "/" instead as of now the service will convert "/" None anyway. I appreciate that is implementation detail that is not exposed and probably shouldn't be baked in to the agent client. Not sure what's preferable there I guess sending root might be better? The path still has to be optional since as of right now we will not be sending back the path per file but instead per package. Which I will address in a further PR.

File: src/bf/model/upload.rs (hunk 3 above)
Comment (ptravers) [reply src/bf/model/upload.rs]: It means root essentially. It means you are not attempting to upload to a collection

**Comment by mattusifer** [comment]
@ptravers Sounds good, thanks for explaining. Yeah I would prefer sending root here especially if the `None`-ness can mean 2 different things (file vs package or root path), and then ideally we would make this non-optional once the file vs package thing stuff is sorted out.

**Comment by mattusifer** [comment]
Ah, that's right. I forgot about that. Yeah I agree we really should split that into separate types at some point...

**Pull request merged**